File: tebd/quntum_quench_pott_chain.jl
using LinearAlgebra
using TensorOperations
using ProgressMeter
using DelimitedFiles
using Plots

###########################################################
# Helpers: NumPy-style matrix <-> tensor map
###########################################################

function mat_to_tensor4(M::AbstractMatrix{<:Complex})
    return permutedims(reshape(M, 3, 3, 3, 3), (2, 1, 4, 3))
end

function tensor4_to_mat(T::AbstractArray{<:Complex,4})
    return reshape(permutedims(T, (2, 1, 4, 3)), 9, 9)
end

###########################################################
# Potts operators
###########################################################

function potts_ops()
    w = exp(2im * π / 3)

    Z = ComplexF64[
        1 0 0
        0 w 0
        0 0 w^2
    ]

    X = ComplexF64[
        0 0 1
        1 0 0
        0 1 0
    ]

    I3 = Matrix{ComplexF64}(I, 3, 3)
    return Z, X, I3
end

###########################################################
# Bond Hamiltonians
###########################################################

function potts_bond_hamiltonian(J, f, L)
    Z, X, I3 = potts_ops()
    Zh = adjoint(Z)
    Xh = adjoint(X)

    H_bond = Vector{Array{ComplexF64,4}}(undef, L - 1)

    for i in 1:(L - 1)
        Hmat = zeros(ComplexF64, 9, 9)

        Hmat .+= -J .* (kron(Zh, Z) + kron(Z, Zh))

        wl = (i == 1)     ? 1.0 : 0.5
        wr = (i == L - 1) ? 1.0 : 0.5

        Hmat .+= -wl * f .* (kron(X, I3) + kron(Xh, I3))
        Hmat .+= -wr * f .* (kron(I3, X) + kron(I3, Xh))

        H_bond[i] = mat_to_tensor4(Hmat)
    end

    return H_bond
end

function potts_expH(J, f, L, dt; imag=false)
    H = potts_bond_hamiltonian(J, f, L)
    gates = Vector{Array{ComplexF64,4}}(undef, length(H))

    for i in eachindex(H)
        hmat = tensor4_to_mat(H[i])
        Umat = imag ? exp(-dt * hmat) : exp(-1im * dt * hmat)
        gates[i] = mat_to_tensor4(Umat)
    end

    return gates
end

###########################################################
# Initial MPS
###########################################################

function initialMPS(L, d)
    G = [zeros(ComplexF64, d, 1, 1) for _ in 1:L]
    l = [ones(Float64, 1) for _ in 1:(L + 1)]

    for i in 1:L
        G[i][1, 1, 1] = 1.0 + 0im
    end

    return l, G
end

function safe_inv(x; eps=1e-14)
    y = zeros(ComplexF64, length(x))
    for i in eachindex(x)
        if abs(x[i]) > eps
            y[i] = 1.0 / x[i]
        end
    end
    return y
end

###########################################################
# Helpers for scaling MPS tensors by Schmidt values
###########################################################

function left_scale_gamma(lvec, G)
    d, χL, χR = size(G)
    out = similar(G)
    for a in 1:χL
        out[:, a, :] .= lvec[a] .* G[:, a, :]
    end
    return out
end

function right_scale_gamma(G, lvec)
    d, χL, χR = size(G)
    out = similar(G)
    for b in 1:χR
        out[:, :, b] .= G[:, :, b] .* lvec[b]
    end
    return out
end

###########################################################
# Two-site TEBD
###########################################################

function tebd_step!(l, G, gates; chi_max=50)
    L = length(G)

    for parity in (0, 1)
        start_idx = (parity == 0) ? 1 : 2

        for i in start_idx:2:(L - 1)
            d    = size(G[i], 1)
            χ0   = size(G[i], 2)
            χ1   = size(G[i], 3)
            χ2   = size(G[i + 1], 3)

            l0 = l[i]
            l1 = l[i + 1]
            l2 = l[i + 2]

            l0inv = safe_inv(l0)
            l2inv = safe_inv(l2)

            # theta[a,s,t,b]
            A = left_scale_gamma(l0, G[i])     # (s,a,c)
            B = right_scale_gamma(A, l1)       # (s,a,c)

            @tensor C[s,t,a,b] := B[s,a,c] * G[i+1][t,c,b]

            theta = Array{ComplexF64}(undef, χ0, d, d, χ2)
            for b in 1:χ2
                theta[:, :, :, b] .= permutedims(C[:, :, :, b], (3, 1, 2)) .* l2[b]
            end

            # apply gate
            # gates[i] has indices (s_out, t_out, s_in, t_in)
            @tensor phi[s,t,a,b] := gates[i][s,t,x,y] * theta[a,x,y,b]

            # reorder to (a,s,t,b), then reshape
            phi = permutedims(phi, (3, 1, 2, 4))
            phi_mat = reshape(phi, χ0 * d, d * χ2)

            F = svd(phi_mat)
            keep = min(length(F.S), chi_max)

            U  = F.U[:, 1:keep]
            S  = F.S[1:keep]
            Vt = F.Vt[1:keep, :]

            nrm = norm(S)
            if nrm > 0
                S ./= nrm
            end

            l[i + 1] = Float64.(S)

            # update G[i]
            Uten = reshape(U, χ0, d, keep)
            Uten = permutedims(Uten, (2, 1, 3))   # (s,a,α)
            for a in 1:χ0
                Uten[:, a, :] .*= l0inv[a]
            end
            G[i] = ComplexF64.(Uten)

            # update G[i+1]
            Vten = reshape(Vt, keep, d, χ2)       # (α,t,b)
            Vten = permutedims(Vten, (2, 1, 3))   # (t,α,b)
            for b in 1:χ2
                Vten[:, :, b] .*= l2inv[b]
            end
            G[i + 1] = ComplexF64.(Vten)
        end
    end

    return nothing
end

###########################################################
# Dense state reconstruction for overlap
# reliable for moderate L, not for huge L
###########################################################

function mps_to_state(l, G)
    psi = ComplexF64[1.0 + 0im;;]  # 1×1 matrix

    for i in 1:length(G)
        psi = psi * Diagonal(ComplexF64.(l[i]))   # (..., χ_i)
        psi = reshape(psi, :, size(psi, 2))

        χleft = size(psi, 2)
        d, χL, χR = size(G[i])
        @assert χleft == χL

        tmp = Array{ComplexF64}(undef, size(psi, 1), d, χR)
        @tensor tmp[a,s,b] := psi[a,c] * G[i][s,c,b]
        psi = reshape(tmp, :, χR)
    end

    psi = psi * Diagonal(ComplexF64.(l[length(G) + 1]))
    return vec(psi)
end

###########################################################
# Efficient overlap
###########################################################

function overlap(l1, G1, l2, G2)
    E = ComplexF64[1.0 + 0im;;]  # 1×1

    L = length(G1)

    for i in 1:L
        A = left_scale_gamma(l1[i], G1[i])   # (d, χL1, χR1)
        B = left_scale_gamma(l2[i], G2[i])   # (d, χL2, χR2)

        # E: (χL1, χL2)
        # A*: (d, χL1, χR1)
        # B : (d, χL2, χR2)

        @tensor Enew[r1,r2] := E[l1i,l2i] * conj(A[s,l1i,r1]) * B[s,l2i,r2]
        E = Enew
    end

    EL = Diagonal(ComplexF64.(l1[L + 1]))
    ER = Diagonal(ComplexF64.(l2[L + 1]))

    val = EL' * E * ER
    return val[1,1]
end

###########################################################
# Main DQPT simulation
###########################################################

function main()
    L = 50
    J = 1.0

    f0 = 10.0
    f1 = 0.0

    dt = 0.02
    steps = 200

    d = 3
    chi_max = 50

    ###########################################################
    # prepare ground state (imaginary time)
    ###########################################################

    l, G = initialMPS(L, d)

    gates0 = potts_expH(J, f0, L, 0.01; imag=true)

    println("Preparing ground state with imaginary time evolution...")
    @showprogress "Imaginary time evolution " for _ in 1:1000
        tebd_step!(l, G, gates0; chi_max=chi_max)
    end

    psi0_l = deepcopy(l)
    psi0_G = deepcopy(G)

    ###########################################################
    # quench evolution
    ###########################################################

    gates1 = potts_expH(J, f1, L, dt; imag=false)

    times = Float64[]
    rate  = Float64[]

    println("Starting quench evolution...")
    @showprogress "Quench evolution " for step in 1:steps
        tebd_step!(l, G, gates1; chi_max=chi_max)

        Gt = overlap(psi0_l, psi0_G, l, G)
        Ival = -(1 / L) * log(abs(Gt)^2)

        push!(times, step * dt * J)
        push!(rate, real(Ival))
    end

    ###########################################################
    # Save data
    ###########################################################

    data = hcat(times, rate)
    filename = "dqpt_potts_L$(L)_f0$(round(f0, digits=2))_f1$(round(f1, digits=2))_julia.dat"

    open(filename, "w") do io
        write(io, "# Jt   I(t)\n")
        writedlm(io, data)
    end

    println("Saved data to: $filename")

    ###########################################################
    # Plot rate function
    ###########################################################

    plt = plot(
        times, rate,
        marker = :circle,
        linewidth = 2,
        xlabel = "J t",
        ylabel = "I(t)",
        title = "DQPT rate function (3-state Potts chain)",
        label = false,
        size = (600, 400)
    )

    display(plt)
end

main()

File: tebd/quntum_quench_pott_chain.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import expm
from numpy.linalg import svd
from tqdm import tqdm
###########################################################
# Potts operators
###########################################################

def potts_ops():
    w = np.exp(2j * np.pi / 3)

    Z = np.array([[1, 0, 0],
                  [0, w, 0],
                  [0, 0, w**2]], dtype=complex)

    X = np.array([[0, 0, 1],
                  [1, 0, 0],
                  [0, 1, 0]], dtype=complex)

    I = np.eye(3, dtype=complex)
    return Z, X, I


###########################################################
# Bond Hamiltonians
###########################################################

def potts_bond_hamiltonian(J, f, L):
    Z, X, I = potts_ops()
    Xh = X.conj().T

    H_bond = []

    for i in range(L - 1):
        H = np.zeros((9, 9), dtype=complex)

        H += -J * (np.kron(Z.conj().T, Z) + np.kron(Z, Z.conj().T))

        wl = 1.0 if i == 0 else 0.5
        wr = 1.0 if i == L - 2 else 0.5

        H += -wl * f * (np.kron(X, I) + np.kron(Xh, I))
        H += -wr * f * (np.kron(I, X) + np.kron(I, Xh))

        H_bond.append(H.reshape(3, 3, 3, 3))

    return H_bond


def potts_expH(J, f, L, dt, imag=False):
    H = potts_bond_hamiltonian(J, f, L)
    gates = []

    for h in H:
        h = h.reshape(9, 9)
        if imag:
            U = expm(-dt * h)
        else:
            U = expm(-1j * dt * h)

        # indices: (s_out, t_out, s_in, t_in)
        gates.append(U.reshape(3, 3, 3, 3))

    return gates


###########################################################
# Initial MPS
###########################################################

def initialMPS(L, d):
    G = []
    l = []

    for _ in range(L):
        G.append(np.zeros((d, 1, 1), dtype=complex))
        l.append(np.ones(1, dtype=complex))

    l.append(np.ones(1, dtype=complex))

    for i in range(L):
        G[i][0, 0, 0] = 1.0

    return l, G


def safe_inv(x, eps=1e-14):
    y = np.zeros_like(x, dtype=complex)
    mask = np.abs(x) > eps
    y[mask] = 1.0 / x[mask]
    return y


###########################################################
# Two-site TEBD
###########################################################

def tebd_step(l, G, gates, chi_max=50):
    L = len(G)

    for parity in [0, 1]:
        for i in range(parity, L - 1, 2):
            d = G[i].shape[0]
            chi0 = G[i].shape[1]
            chi1 = G[i].shape[2]
            chi2 = G[i + 1].shape[2]

            l0 = l[i]
            l1 = l[i + 1]
            l2 = l[i + 2]

            l0inv = safe_inv(l0)
            l2inv = safe_inv(l2)

            theta = np.tensordot(np.diag(l0), G[i], axes=(1, 1))
            theta = np.tensordot(theta, np.diag(l1), axes=(2, 0))
            theta = np.tensordot(theta, G[i + 1], axes=(2, 1))
            theta = np.tensordot(theta, np.diag(l2), axes=(3, 0))
            # shape: (chi0, d, d, chi2)

            # correct gate application
            theta = np.tensordot(gates[i], theta, axes=([2, 3], [1, 2]))
            theta = np.transpose(theta, (2, 0, 1, 3))
            # shape: (chi0, d, d, chi2)

            theta = theta.reshape(chi0 * d, d * chi2)

            U, S, Vh = svd(theta, full_matrices=False)

            keep = min(len(S), chi_max)

            U = U[:, :keep]
            S = S[:keep]
            Vh = Vh[:keep, :]

            nrm = np.linalg.norm(S)
            if nrm > 0:
                S = S / nrm

            l[i + 1] = S

            U = U.reshape(chi0, d, keep)
            U = np.tensordot(np.diag(l0inv), U, axes=(1, 0))
            U = np.transpose(U, (1, 0, 2))
            G[i] = U

            V = Vh.reshape(keep, d, chi2)
            V = np.transpose(V, (1, 0, 2))
            V = np.tensordot(V, np.diag(l2inv), axes=(2, 0))
            G[i + 1] = V


###########################################################
# Dense state reconstruction for overlap
# reliable for moderate L, not for huge L
###########################################################

def mps_to_state(l, G):
    psi = np.array([[1.0 + 0.0j]])  # (1, chi_left)

    for i in range(len(G)):
        psi = np.tensordot(psi, np.diag(l[i]), axes=(1, 0))   # (..., chi_i)
        psi = np.tensordot(psi, G[i], axes=(1, 1))            # (..., d, chi_{i+1})

    psi = np.tensordot(psi, np.diag(l[len(G)]), axes=(psi.ndim - 1, 0))
    psi = np.squeeze(psi, axis=(-1, 0))
    return psi.reshape(-1)


# def overlap(l1, G1, l2, G2):
#     psi1 = mps_to_state(l1, G1)
#     psi2 = mps_to_state(l2, G2)
#     return np.vdot(psi1, psi2)

def overlap(l1, G1, l2, G2):
    E = np.array([[1.0 + 0.0j]])

    L = len(G1)

    for i in range(L):
        A = np.tensordot(np.diag(l1[i]), G1[i], axes=(1, 1))          # (chiL1,d,chiR1)
        B = np.tensordot(np.diag(l2[i]), G2[i], axes=(1, 1))          # (chiL2,d,chiR2)

        E = np.tensordot(E, A.conj(), axes=(0, 0))                    # (chiL2,d,chiR1)
        E = np.tensordot(E, B, axes=([0, 1], [0, 1]))                 # (chiR1,chiR2)

    E = np.tensordot(E, np.diag(l1[L]).conj(), axes=(0, 0))
    E = np.tensordot(E, np.diag(l2[L]), axes=([0, 1], [0, 1]))

    return E.reshape(())
###########################################################
# Main DQPT simulation
###########################################################

L = 50
J = 1.0

f0 = 10.0
f1 = 0.0

dt = 0.02
steps = 200

d = 3
chi_max = 50

###########################################################
# prepare ground state (imaginary time)
###########################################################

l, G = initialMPS(L, d)

gates0 = potts_expH(J, f0, L, 0.01, imag=True)

print("Preparing ground state with imaginary time evolution...")
# use tqdm for progress bar
for _ in tqdm(range(1000), desc="Imaginary time evolution"):
    tebd_step(l, G, gates0, chi_max=chi_max)

psi0_l = [x.copy() for x in l]
psi0_G = [x.copy() for x in G]

###########################################################
# quench evolution
###########################################################

gates1 = potts_expH(J, f1, L, dt, imag=False)

times = []
rate = []

# use tqdm for progress bar
print("Starting quench evolution...")
for step in tqdm(range(steps), desc="Quench evolution"):
    tebd_step(l, G, gates1, chi_max=chi_max)

    Gt = overlap(psi0_l, psi0_G, l, G)
    I = -(1 / L) * np.log(np.abs(Gt) ** 2)

    times.append((step + 1) * dt * J)
    rate.append(I.real)

###########################################################
# Save data
###########################################################

data = np.column_stack((times, rate))
np.savetxt(
    "dqpt_potts_L%d_f0%.2f_f1%.2f.dat" % (L, f0, f1),
    data,
    header="Jt   I(t)"
)

###########################################################
# Plot rate function
###########################################################

plt.figure(figsize=(6, 4))
plt.plot(times, rate, 'o-')
plt.xlabel("J t")
plt.ylabel("I(t)")
plt.title("DQPT rate function (3-state Potts chain)")
plt.tight_layout()
plt.show()